File: android/app/src/main/java/com/fmaldonado/siase/data/repositories/AuthRepository.kt
package com.fmaldonado.siase.data.repositories

import android.util.Log
import com.fmaldonado.siase.data.models.Preferences
import com.fmaldonado.siase.data.models.SignInRequest
import com.fmaldonado.siase.data.models.SignInResponse
import com.fmaldonado.siase.data.network.NetworkDataSource
import com.fmaldonado.siase.data.persistence.dao.MainCareerDao
import com.fmaldonado.siase.data.persistence.dao.MainScheduleDao
import com.fmaldonado.siase.data.preferences.PreferencesService
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository
@Inject
constructor(
    private val networkDataSource: NetworkDataSource,
    private val preferencesService: PreferencesService,
    private val mainCareerDao: MainCareerDao,
    private val mainScheduleDao: MainScheduleDao
) {

    var signedInUser: SignInResponse?
        private set(value) {
            val preferences = preferencesService.getPreferences()
            preferences.session = value
            preferencesService.save(preferences)
        }
        get() {
            val preferences = preferencesService.getPreferences()
            return preferences.session
        }


    fun getSession(): Preferences {
        return preferencesService.getPreferences()
    }

    fun checkSession(): Boolean {
        val preferences = preferencesService.getPreferences()

        if (preferences.user == null ||
            preferences.password == null
        )
            return false

        return true
    }

    suspend fun restoreSession() {
        val preferences = preferencesService.getPreferences()
        signIn(preferences.user!!, preferences.password!!)
    }

    suspend fun signIn(username: String, password: String): SignInResponse? {
        val result = networkDataSource.signIn(SignInRequest(username, password))

        val preferences = preferencesService.getPreferences()

        preferences.session = result
        preferences.user = username
        preferences.password = password

        preferencesService.save(preferences)

        return signedInUser
    }

    suspend fun signOut() {
        preferencesService.delete()
        mainCareerDao.deleteMainCareer()
        mainScheduleDao.deleteMainSchedule()
    }

}

File: ios/siase/data/repositories/AuthRepository.swift
//
//  AuthRepository.swift
//  siase
//
//  Created by Fernando Maldonado on 27/02/22.
//

import Foundation


class AuthRepository{
    
    private let networkDataSource:NetworkDataSource
    
    private let preferencesService:PreferencesService
    private let mainCareerDao:MainCareerDao
    private let mainScheduleDao:MainScheduleDao
    private let mainSheduleClassesDao:MainScheduleClassesDao
    
    var currentSession:LoginResponse?
    {
        set{
            var preferences = self.preferencesService.getPreferences()
            preferences.session = newValue
            self.preferencesService.savePreferences(newPreferences: preferences)
        }
        get{
            self.preferencesService.getPreferences().session
        }
    }
    
    
    
    init(
        networkDataSource:NetworkDataSource =
        DIContainer.shared.resolve(type: NetworkDataSource.self)!,
        preferencesService:PreferencesService =
        DIContainer.shared.resolve(type: PreferencesService.self)!,
        mainCareerDao:MainCareerDao =
        DIContainer.shared.resolve(type: MainCareerDao.self)!,
        mainScheduleDao:MainScheduleDao =
        DIContainer.shared.resolve(type: MainScheduleDao.self)!,
        mainScheduleClassesDao:MainScheduleClassesDao =
        DIContainer.shared.resolve(type: MainScheduleClassesDao.self)!
        
    ){
        self.networkDataSource = networkDataSource
        self.preferencesService = preferencesService
        self.mainCareerDao = mainCareerDao
        self.mainScheduleDao = mainScheduleDao
        self.mainSheduleClassesDao = mainScheduleClassesDao
    }
    
    func checkSession() -> Bool{
        let preferences = preferencesService.getPreferences()
        
        
        if(preferences.user == nil || preferences.password == nil){
            return false
        }
        
        return true
    }
    
    func restoreSession(completer:@escaping (LoginResponse?,AppError?)->Void){
        let preferences = preferencesService.getPreferences()
        self.login(
            username: preferences.user!,
            password: preferences.password!,
            completer: completer
        )
    }
    
    func login(
        username:String,
        password:String,
        completer:@escaping (LoginResponse?,AppError?)->Void
    ){
        networkDataSource.login(
            user: username,
            password: password
        ){ response,error in
            
            if(error != nil){
                return completer(nil,error)
            }
            
            var preferences = Preferences()
            preferences.user = username
            preferences.password = password
            preferences.session = response
            self.preferencesService.savePreferences(newPreferences: preferences)
            
            completer(response,nil)
        }
    }
    
    func signOut(){
        self.preferencesService.delete()
        try? self.mainCareerDao.deleteMainCareer()
        try? self.mainScheduleDao.deleteMainCareer()
        try? self.mainSheduleClassesDao.deleteClasses()
    }
}
